File: ios/TmuxWebClient/Views/Changes/FileTree.swift
import Foundation

/// One node in the folder tree `buildFileTree` groups a flat `ChangedFile`
/// list into. `file` is set only on leaf (file) nodes; folder nodes carry
/// `nil` and derive their children from every path that shares their
/// prefix. Mirrors `buildFileTree`/`renderTreeChildren` in
/// ../../../public/app.js -- folders sort before files, both
/// alphabetically.
struct FileTreeNode: Identifiable, Hashable {
    let name: String
    let children: [FileTreeNode]
    let file: ChangedFile?

    var id: String { file?.id ?? name }
    var isFolder: Bool { file == nil }
}

/// Pure function (no SwiftUI dependency) so tree-grouping logic is
/// testable in isolation from the rendering layer -- see
/// ChangesListView.swift for the consumer.
func buildFileTree(_ files: [ChangedFile]) -> [FileTreeNode] {
    let entries = files
        .map { (path: $0.path.split(separator: "/").map(String.init), file: $0) }
        // A path that splits to an empty array (e.g. an empty string) would
        // crash buildLevel's `$0.path[0]` below -- drop it defensively since
        // an empty path can't be rendered as a tree node anyway. Not
        // currently reachable via `git status --porcelain` (paths are never
        // empty there), but ChangedFile.path is server-decoded JSON with no
        // type-level non-emptiness guarantee.
        .filter { !$0.path.isEmpty }
    return buildLevel(entries)
}

private func buildLevel(_ entries: [(path: [String], file: ChangedFile)]) -> [FileTreeNode] {
    let grouped = Dictionary(grouping: entries, by: { $0.path[0] })

    let nodes = grouped.map { name, group -> FileTreeNode in
        let leaf = group.first { $0.path.count == 1 }
        let deeper = group
            .filter { $0.path.count > 1 }
            .map { (path: Array($0.path.dropFirst()), file: $0.file) }

        return FileTreeNode(name: name, children: buildLevel(deeper), file: leaf?.file)
    }

    return nodes.sorted { lhs, rhs in
        if lhs.isFolder != rhs.isFolder { return lhs.isFolder }
        return lhs.name.localizedStandardCompare(rhs.name) == .orderedAscending
    }
}

File: web/src/domain/fileTree.ts
// Ports kmp/composeApp/.../domain/FileTreeNode.kt -- groups a flat
// ChangedFile list into a folder tree. Mirrors `buildFileTree`/
// `renderTreeChildren` in public/app.js and FileTree.swift: folders sort
// before files, both alphabetically.
import type { ChangedFile } from "../api/types";

export interface FileTreeNode {
  name: string;
  children: FileTreeNode[];
  file: ChangedFile | null;
  id: string;
  isFolder: boolean;
}

function makeNode(name: string, children: FileTreeNode[], file: ChangedFile | null): FileTreeNode {
  return { name, children, file, id: file?.path ?? name, isFolder: file === null };
}

/** Pure function so tree-grouping logic is testable in isolation from any UI layer. */
export function buildFileTree(files: ChangedFile[]): FileTreeNode[] {
  const entries = files
    .map((file): [string[], ChangedFile] => [file.path.split("/").filter((segment) => segment.length > 0), file])
    // A path that splits to an empty list (e.g. an empty string) can't be
    // rendered as a tree node -- dropped defensively. Not currently
    // reachable via `git status --porcelain` (paths are never empty
    // there), but ChangedFile.path is server-decoded JSON with no
    // type-level non-emptiness guarantee.
    .filter(([segments]) => segments.length > 0);
  return buildLevel(entries);
}

function buildLevel(entries: [string[], ChangedFile][]): FileTreeNode[] {
  const grouped = new Map<string, [string[], ChangedFile][]>();
  for (const entry of entries) {
    const [segments] = entry;
    const name = segments[0];
    const group = grouped.get(name);
    if (group) group.push(entry);
    else grouped.set(name, [entry]);
  }

  const nodes = Array.from(grouped.entries()).map(([name, group]) => {
    const leaf = group.find(([segments]) => segments.length === 1);
    const deeper: [string[], ChangedFile][] = group
      .filter(([segments]) => segments.length > 1)
      .map(([segments, file]) => [segments.slice(1), file]);

    return makeNode(name, buildLevel(deeper), leaf ? leaf[1] : null);
  });

  return nodes.sort((a, b) => {
    if (a.isFolder !== b.isFolder) return a.isFolder ? -1 : 1;
    // Ordinal comparison (not localeCompare) to match Kotlin's String.compareTo exactly.
    return a.name < b.name ? -1 : a.name > b.name ? 1 : 0;
  });
}
